async-nats/src/jetstream/kv/mod.rs

```diff
@@ -0,0 +1,499 @@
+use time::OffsetDateTime;
+
+use crate::{header, jetstream::response, Error, Message};
+
+use self::bucket::Status;
+
+use super::{
+    consumer::DeliverPolicy,
+    stream::{RawMessage, StorageType, Stream},
+};
+
+// Helper to extract key value operation from message headers
+fn kv_operation_from_maybe_headers(maybe_headers: Option<&String>) -> Operation {
+    if let Some(headers) = maybe_headers {
+        return match headers.as_str() {
+            KV_OPERATION_DELETE => Operation::Delete,
+            KV_OPERATION_PURGE => Operation::Purge,
+            _ => Operation::Put,
+        };
+    }
+
+    Operation::Put
+}
+
+fn kv_operation_from_stream_message(message: &RawMessage) -> Operation {
+    kv_operation_from_maybe_headers(message.headers.as_ref())
+}
+lazy_static! {
+    static ref VALID_BUCKET_RE: Regex = Regex::new(r#"\A[a-zA-Z0-9_-]+\z"#).unwrap();
+    static ref VALID_KEY_RE: Regex = Regex::new(r#"\A[-/_=\.a-zA-Z0-9]+\z"#).unwrap();
+}
+
+pub(crate) const MAX_HISTORY: i64 = 64;
+const ALL_KEYS: &str = ">";
+
+const KV_OPERATION: &str = "KV-Operation";
+const KV_OPERATION_DELETE: &str = "DEL";
+const KV_OPERATION_PURGE: &str = "PURGE";
+const KV_OPERATION_PUT: &str = "PUT";
+
+const NATS_ROLLUP: &str = "Nats-Rollup";
+const ROLLUP_SUBJECT: &str = "sub";
+
+pub(crate) fn is_valid_bucket_name(bucket_name: &str) -> bool {
+    VALID_BUCKET_RE.is_match(bucket_name)
+}
+
+pub(crate) fn is_valid_key(key: &str) -> bool {
+    if key.is_empty() || key.starts_with('.') || key.ends_with('.') {
+        return false;
+    }
+
+    VALID_KEY_RE.is_match(key)
+}
+
+/// Configuration values for key value stores.
+#[derive(Debug, Default)]
+pub struct Config {
+    /// Name of the bucket
+    pub bucket: String,
+    /// Human readable description.
+    pub description: String,
+    /// Maximum size of a single value.
+    pub max_value_size: i32,
+    /// Maximum historical entries.
+    pub history: i64,
+    /// Maximum age of any entry in the bucket, expressed in nanoseconds
+    pub max_age: std::time::Duration,
+    /// How large the bucket may become in total bytes before the configured discard policy kicks in
+    pub max_bytes: i64,
+    /// The type of storage backend, `File` (default) and `Memory`
+    pub storage: StorageType,
+    /// How many replicas to keep for each entry in a cluster.
+    pub num_replicas: usize,
+    // TODO: add placement
+}
+
+/// Describes what kind of operation and entry represents
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Operation {
+    /// A value was put into the bucket
+    Put,
+    /// A value was deleted from a bucket
+    Delete,
+    /// A value was purged from a bucket
+    Purge,
+}
+
+#[derive(Debug, Clone)]
+pub struct Store {
+    pub name: String,
+    pub stream_name: String,
+    pub prefix: String,
+    pub stream: Stream,
+}
+
+impl Store {
+    pub async fn status(&self) -> Result<Status, Error> {
+        // TODO: should we poll for fresh info here? probably yes.
```

Comment:
Probably the same as streams etc, we have a cached and query version?

Reply:
we have cached and updating version.

the latter is `&mut self`. I can make it `query_status`  and pub.